ddtrace/contrib/grpc/patch.py

```diff
@@ -1,64 +1,85 @@
 import grpc
-from ddtrace.vendor import wrapt
 
-from ddtrace import Pin
-from ...utils.wrappers import unwrap
+from ddtrace.vendor.wrapt import wrap_function_wrapper as _w
+from ddtrace import config, Pin
 
-from .client_interceptor import GrpcClientInterceptor
+from ...utils.wrappers import unwrap as _u
+
+from .client_interceptor import create_client_interceptor
+from .server_interceptor import create_server_interceptor
+
+
+config._add('grpc', dict(
+    service_name='grpc',
+    distributed_tracing_enabled=True,
+))
 
 
 def patch():
     # patch only once
     if getattr(grpc, '__datadog_patch', False):
         return
     setattr(grpc, '__datadog_patch', True)
-    Pin(service='grpc', app='grpc', app_type='grpc').onto(grpc)
 
-    _w = wrapt.wrap_function_wrapper
+    Pin(service=config.grpc.service_name).onto(grpc)
 
-    _w('grpc', 'insecure_channel', _insecure_channel_with_interceptor)
-    _w('grpc', 'secure_channel', _secure_channel_with_interceptor)
+    _w('grpc', 'insecure_channel', _client_channel_interceptor)
+    _w('grpc', 'secure_channel', _client_channel_interceptor)
+    _w('grpc', 'server', _server_constructor_interceptor)
 
 
 def unpatch():
     if not getattr(grpc, '__datadog_patch', False):
         return
     setattr(grpc, '__datadog_patch', False)
-    unwrap(grpc, 'secure_channel')
-    unwrap(grpc, 'insecure_channel')
 
+    pin = Pin.get_from(grpc)
+    if pin:
+        pin.remove_from(grpc)
 
-def _insecure_channel_with_interceptor(wrapped, instance, args, kwargs):
+    _u(grpc, 'secure_channel')
+    _u(grpc, 'insecure_channel')
+    _u(grpc, 'server')
+
+
+def _client_channel_interceptor(wrapped, instance, args, kwargs):
     channel = wrapped(*args, **kwargs)
 
-    if 'target' in kwargs:
-        target = kwargs['target']
-    else:
-        target = args[0]
+    (host, port) = _parse_target_from_arguments(args, kwargs)
 
-    (host, port) = get_host_port(target)
-    channel = _intercept_channel(channel, host, port)
-    return channel
+    pin = Pin.get_from(grpc)
+    if not pin:
+        return channel
 
+    interceptor_function = create_client_interceptor(pin, host, port)
+    return grpc.intercept_channel(channel, interceptor_function)
 
-def _secure_channel_with_interceptor(wrapped, instance, args, kwargs):
-    channel = wrapped(*args, **kwargs)
 
-    if 'target' in kwargs:
-        target = kwargs['target']
-    else:
-        target = args[0]
+def _server_constructor_interceptor(wrapped, instance, args, kwargs):
+    # DEV: we clone the pin on the grpc module and configure it for the server
+    # interceptor
+
+    pin = Pin.get_from(grpc)
+    if not pin:
+        return wrapped(*args, **kwargs)
+
+    interceptor = create_server_interceptor(pin)
 
-    (host, port) = get_host_port(target)
-    channel = _intercept_channel(channel, host, port)
-    return channel
+    # DEV: Inject our tracing interceptor first in the list of interceptors
+    if 'interceptors' in kwargs:
```

Comment:
is it possible for it to be provided as `args`?

e.g. do we want:

```python
if len(args) > 4:
    args[3] = (interceptor ,) + tuple(args[3])
elif 'interceptors' in kwargs:
    pass
else:
    pass
```

Reply:
This is the method we are wrapping and specifies `interceptor` as a keyword agrument:

https://github.com/grpc/grpc/blob//src/python/grpcio/grpc/__init__.py#L1638